Spring Boot to management collections of vocabularies

vukhuongduy123 opened the pull request.

vukhuongduy123 committed:
Spring Boot to management collections of vocabularies

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,72 @@
+public class CollectionController {
+	private final CollectionService collectionService;
+
+	@GetMapping("collections")
```
put "collections" to line 25

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/VocabularyController.java:
```diff
@@ -0,0 +1,73 @@
+public class VocabularyController {
+	private final VocabularyService vocabularyService;
+
+	@GetMapping("collections/{collection-id}/vocabularies")
```
better to move it to Collection controller

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/VocabularyController.java:
```diff
@@ -0,0 +1,73 @@
+	@PreAuthorize("hasAnyAuthority('Admin')")
+	public ResponseEntity<Void> save(@PathVariable("collection-id") UUID collectionId, @RequestBody VocabularyCreateRequest request) {
+		vocabularyService.save(request, collectionId);
+		return ResponseEntity.ok().build();
```
wwe should return the saved itself. Client will make use of the id of it

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,72 @@
+	@PreAuthorize("hasAnyAuthority('Admin')")
+	public ResponseEntity<Void> save(@RequestBody CollectionCreateRequest request) {
+		collectionService.save(request, ContextUtils.extractUserIdFromSecurityContext());
+		return ResponseEntity.ok().build();
```
wwe should return the saved itself. Client will make use of the id of it

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,72 @@
+	@PreAuthorize("hasAnyAuthority('Admin')")
+	public ResponseEntity<Void> update(@RequestBody CollectionUpdateRequest request) {
+		collectionService.update(request);
+		return ResponseEntity.ok().build();
```
wwe should return the saved itself. Client will check whether the updated data is correct or not

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/VocabularyController.java:
```diff
@@ -0,0 +1,73 @@
+	@PreAuthorize("hasAnyAuthority('Admin')")
+	public ResponseEntity<Void> update(@RequestBody VocabularyUpdateRequest request) {
+		vocabularyService.update(request);
+		return ResponseEntity.ok().build();
```
wwe should return the saved itself. Client will check whether the updated data is correct or not

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/WordController.java:
```diff
@@ -0,0 +1,71 @@
+
+	@GetMapping("vocabularies/{vocabulary-id}/words")
+	@PreAuthorize("hasAnyAuthority('Viewer', 'Admin')")
+	public ResponseEntity<PageResponse<WordResponse>> findAll(@PathVariable("vocabulary-id") UUID vocabularyId,
```
better to move to vocab controller

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/WordController.java:
```diff
@@ -0,0 +1,71 @@
+		return ResponseEntity.ok(wordService.findById(wordId));
+	}
+
+	@PostMapping("vocabularies/{vocabulary-id}/words")
```
same

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/WordController.java:
```diff
@@ -0,0 +1,71 @@
+		return ResponseEntity.ok(wordService.findAll(pageNumber, pageSize, vocabularyId));
+	}
+
+	@GetMapping("words/{word-id}")
```
move all "words" to line 23

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/service/impl/CollectionServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+	public CollectionResponse findById(UUID id) {
+		return CollectionResponseMapper.INSTANCE.map(
+				collectionRepository.findById(id)
+									.orElseThrow(() -> new InvalidRequestParamException("Collection with id " + id + "not found ")));
```
should be ResourceNotFOundException

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/service/impl/CollectionServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+	@Transactional
+	public void save(CollectionCreateRequest request, UUID userId) {
+		Collection collection = CollectionRequestMapper.INSTANCE.map(request);
+		collection.setUserId(userId.toString());
```
you dont need to pass user id to the method. use Spring Auditing and the field will be set by SPring

PaulNgo-BlueOC commented on src/main/resources/db/changelog/2024/09/22-01-changelog.xml:
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
how do you gen these change logs?

vukhuongduy123 replied:
I generated these change logs by liquibase plugin in intelliJ based on entities

vukhuongduy123 committed:
- User Spring AuditorAware to auto include user_id in collection
- Move controller methods
- Return saved id for update/insert

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	@GetMapping("/name")
+	@PreAuthorize("hasAnyAuthority('Viewer', 'Admin')")
```
it should be merge to `findAll()`,  change `name` to `keyword`, use reqeust param

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,100 @@
+
+	@GetMapping("/{collection-id}/vocabularies/name")
+	@PreAuthorize("hasAnyAuthority('Viewer', 'Admin')")
+	public ResponseEntity<PageResponse<VocabularyResponse>> findAllVocabulariesInCollectionByName(
```
change `name` to `keyword`, use reqeust param

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,100 @@
+
+	@GetMapping
+	@PreAuthorize("hasAnyAuthority('Viewer', 'Admin')")
+	public ResponseEntity<PageResponse<CollectionResponse>> findAll(@RequestParam(name = "page-size") int pageSize,
```
no need to manually define pageSize, pageNumber. Use Pageable from Spring

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/VocabularyController.java:
```diff
@@ -0,0 +1,76 @@
+
+	@GetMapping("/{vocabulary-id}/words/content")
+	@PreAuthorize("hasAnyAuthority('Viewer', 'Admin')")
+	public ResponseEntity<PageResponse<WordResponse>> findAllWordsInVocabularyByContent(@PathVariable("vocabulary-id") UUID vocabularyId,
```
change `content` to `keyword`, use reqeust param

PaulNgo-BlueOC commented:
All controllers have common methods: findAll(keyword), findById(UUID), save, update, delete. The same for all services. Try to genericize them

vukhuongduy123 committed:
- Merge methods in controller allow to query by param optionally
- Use Specification to build dynamic query based on optional input params(nullable)

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/controller/CollectionController.java:
```diff
@@ -0,0 +1,81 @@
+@RestController
+@RequestMapping("collections")
+@RequiredArgsConstructor
+public class CollectionController {
```
```suggestion
public class CollectionController implements/extends BaseCrudResource<CollectionCreateRequest, CollectionResponse> {
```

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/model/request/CollectionCreateRequest.java:
```diff
@@ -0,0 +1,19 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CollectionCreateRequest {
```
```suggestion
public class CollectionRequest {
```

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/model/request/CollectionUpdateRequest.java:
```diff
@@ -0,0 +1,23 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CollectionUpdateRequest {
```
```suggestion
public class CollectionRequest {
```

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/utils/specification/SpecificationBuilder.java:
```diff
@@ -0,0 +1,37 @@
+		};
+	}
+
+	public static <T, E, X> Specification<T> equalManyToOne(SingularAttribute<T, E> path, SingularAttribute<E, ?> idField, @Nullable X e) {
```
```suggestion
	public static <T, E, X> Specification<T> equalManyToOne(@Nullable X e, SingularAttribute<T, E/?>... path) {
```

PaulNgo-BlueOC commented on src/main/java/org/example/collection/vocabulary/utils/ContextUtils.java:
```diff
@@ -0,0 +1,20 @@
+		if (authentication instanceof JwtAuthenticationToken token) {
+			return UUID.fromString(token.getName());
+		}
+		return null;
```
QUĂNG EX

vukhuongduy123 committed:
- Add a generic type for similar controllers and services

vukhuongduy123 committed:
add mapping methods
